File: src/main/javabak/materials/MaterialBlock.java
package redcrafter07.processed.materials;

import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour;

public abstract class MaterialBlock extends Block {
    protected final Material material;

    public MaterialBlock(Material material, BlockBehaviour.Properties properties) {
        super(properties);
        this.material = material;
    }

    public Material getMaterial() {
        return material;
    }

    public static class MetalBlock extends MaterialBlock {
        public MetalBlock(Material material) {
            super(material, material.getMetalBlockProperties());
        }
    }

    public static class OreBlock extends MaterialBlock {
        public OreBlock(Material material) {
            super(material, material.getOreBlockProperties());
        }
    }
}

File: src/main/kotlin/redcrafter07/processed/materials/MaterialBlock.kt
package redcrafter07.processed.materials

import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import redcrafter07.processed.materials.data.MaterialBase

abstract class MaterialBlock(override val material: MaterialBase, itemProperties: Properties) : Block(itemProperties),
    MaterialContainer {
    class OreBlock(material: MaterialBase) : MaterialBlock(material, Properties.ofFullCopy(Blocks.IRON_ORE))
}